### PR TITLE
Upgrade script: check that installed apps are compatible with the future version before actually starting the upgrade

### PR DESCRIPTION
**Pull request by alexAubin**
## Problem

- cf the classic issue of "Upgrade miserably fails" ... *share log* ... "Yes, app foobar is not compatible ..."
- fix https://github.com/YunoHost-Apps/nextcloud_ynh/issues/590 i guess ?

## Solution

- Implement a check that the installed apps are compatible ... this is not trivial : 
   - gotta list the installed apps (this part is mostly straightforward)
   - substract the "core" apps which are always compatible ... listed in https://github.com/nextcloud/server/blob/master/core/shipped.json
   - substract the apps that will indeed be available in the future version ... listed in https://apps.nextcloud.com/api/v1/platform/28.0.0/apps.json (for 28.x)

## PR Status

- [ ] Code finished and ready to be reviewed/tested
- [ ] The fix/enhancement were manually tested (if applicable)

## Automatic tests

Automatic tests can be triggered on https://ci-apps-dev.yunohost.org/ *after creating the PR*, by commenting "!testme", "!gogogadgetoci" or "By the power of systemd, I invoke The Great App CI to test this Pull Request!". (N.B. : for this to work you need to be a member of the Yunohost-Apps organization)

### CODE REVIEW

**Comment by ericgaspar** [comment]
!testme

File: scripts/upgrade [hunk 1]
```diff
@@ -74,16 +74,45 @@ local mount_id=$(exec_occ files_external:create --output=json \
...
+    local nextcloud_destination_version="$1"
+    local nextcloud_current_version="$(grep OC_VersionString "$install_dir/version.php" | cut -d\' -f2)"
+    local installed_apps=$(mktemp)
+    local core_apps_in_current_version=$(mktemp)
+    local nextcloud_destination_appcatalog=$(mktemp)
+
+    # List installed apps
+    exec_occ app:list --output json | jq -r ".enabled | keys[]" | sort > $installed_apps
+    # Fetch Nextcloud list of core apps from their github repo for the current version
+    curl -s https://raw.githubusercontent.com/nextcloud/server/v$nextcloud_current_version.0.0/core/shipped.json | jq -r '.shippedApps[]' | sort > $core_apps_in_current_version
```
Comment (alexAubin) [inline scripts/upgrade:88]:
```suggestion
    curl -s https://raw.githubusercontent.com/nextcloud/server/v$nextcloud_current_version/core/shipped.json | jq -r '.shippedApps[]' | sort > $core_apps_in_current_version
```

File: scripts/upgrade [hunk 2]
```diff
@@ -74,16 +74,45 @@ local mount_id=$(exec_occ files_external:create --output=json \
...
+    local installed_apps=$(mktemp)
+    local core_apps_in_current_version=$(mktemp)
+    local nextcloud_destination_appcatalog=$(mktemp)
+
+    # List installed apps
+    exec_occ app:list --output json | jq -r ".enabled | keys[]" | sort > $installed_apps
+    # Fetch Nextcloud list of core apps from their github repo for the current version
+    curl -s https://raw.githubusercontent.com/nextcloud/server/v$nextcloud_current_version/core/shipped.json | jq -r '.shippedApps[]' | sort > $core_apps_in_current_version
+    # Fetch Nextcloud app catalog (doesnt contain core app) for the future version
+    curl -s https://apps.nextcloud.com/api/v1/platform/$nextcloud_destination_appcatalog.0.0/apps.json | jq -r '.[] | .id' | sort > $nextcloud_destination_appcatalog
```
Comment (alexAubin) [inline scripts/upgrade:90]:
```suggestion
    curl -s https://apps.nextcloud.com/api/v1/platform/$nextcloud_destination_version.0.0/apps.json | jq -r '.[] | .id' | sort > $nextcloud_destination_appcatalog
```

**Comment by alexAubin** [comment]
!testme

**Comment by alexAubin** [comment]
!testme dude

**Comment by alexAubin** [comment]
(Pending job : https://ci-apps-dev.yunohost.org/ci/job/12741 )

**Comment by alexAubin** [comment]
!testme

**Comment by alexAubin** [comment]
!testme

**Comment by alexAubin** [comment]
Hell yes it's working

**Pull request merged**